com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationArchiveExtractor.java

```diff
@@ -0,0 +1,200 @@
+
+    static final String PATH_SHOULD_NOT_BE_ABSOLUTE = "Archive entry name \"{0}\" should not be absolute";
+    static final String PATH_SHOULD_BE_NORMALIZED = "Archive entry name \"{0}\" should be normalized";
+    private static final int BUFFER_SIZE = 4 * 1024; // 4KB
+    private static final String ARCHIVE_FILE_SEPARATOR = "/";
+    private InputStream appArchiveStream;
+    private String fileName;
+
+    public ApplicationArchiveExtractor(InputStream appArchiveStream, String fileName) {
+        this.appArchiveStream = appArchiveStream;
+        this.fileName = fileName;
+    }
+
+    public Path extract(ExtractStatusCallback callback) throws IOException, SLException {
+        if (callback == null) {
+            callback = ExtractStatusCallback.NONE;
+        }
+
+        try (InputStreamProducer streamProducer = new InputStreamProducer(appArchiveStream, fileName)) {
+            return extract(streamProducer, callback);
+        } catch (IOException | SLException e) {
+            callback.onError(e);
+            throw e;
+        } catch (Exception e) {
+            callback.onError(e);
+            return null;
+        }
+    }
+
+    private Path extract(InputStreamProducer streamProducer, ExtractStatusCallback callback) throws IOException, SLException {
+        InputStream stream = streamProducer.getNextInputStream();
+        if (stream == null) {
+            throw new ContentException(com.sap.cloud.lm.sl.mta.message.Messages.CANNOT_FIND_ARCHIVE_ENTRY, streamProducer.getFileName());
+        }
+
+        if (isFile(streamProducer.getFileName())) {
+            return saveToFile(streamProducer, callback);
+        }
+
+        if (streamProducer.getFileName()
+            .equals(streamProducer.getStreamEntryName())) {
+            return saveToDirectory(streamProducer, callback);
+        }
+
+        return saveToDirectoryUnordered(streamProducer, callback);
+    }
+
+    private Path saveToFile(InputStreamProducer streamProducer, ExtractStatusCallback callback) throws IOException, SLException {
+        String prefix = getPrefix(streamProducer.getStreamEntryName());
+        String suffix = getSuffix(streamProducer.getStreamEntryName());
+
+        Path filePath = File.createTempFile(prefix, suffix)
+            .toPath();
+        callback.onFileCreated(filePath);
+        LOGGER.debug(MessageFormat.format("Saving input stream to temporary file \"{0}\"...", filePath));
+        try (OutputStream outputStream = Files.newOutputStream(filePath)) {
+            copy(streamProducer.getInputStream(), outputStream, callback);
+        }
+        LOGGER.debug(MessageFormat.format("Input stream saved to temporary file \"{0}\"", filePath));
+        return filePath;
+    }
+
+    private Path saveToDirectory(InputStreamProducer streamProducer, ExtractStatusCallback callback) throws IOException, SLException {
+        Path dirPath = getTempDirectory(streamProducer.getFileName());
+        callback.onFileCreated(dirPath);
```

Comment:
The method name here is a bit misleading. It is called "onFileCreated", but here you're giving a directory instead of a file.

Reply:
If the application is a single file (not a directory), then onFileCreated will be called with the path to the file (this case is handled by saveToFile). Accordingly, same will happen if it is not a single file but directory with multiple files. It will return the path to the root folder (which is a file by itself). I do not find any other name more appropriate?